Feature/2019 07 11 js testing 101 (#521)
ordina-jworks/ordina-jworks.github.io: _posts/2019-07-11-JS-testing-101.md

﻿---
layout: post
authors: [ivar_van_hartingsveldt]

> title: 'Javascript Testing 101'

Reviewer: ```suggestion
title: 'JavaScript Testing 101'
```

image: /img/2019-07-11-js-testing.png

> tags: [Spring IO,Spring,Java,Conference]

Reviewer 2: Please update your tags :-)
tags: [JavaScript,Testing,TypeScript,Jasmine,Karma,Unit testing,Observables]

> category: Javascript, Testing, Typescript, Jasmine, Karma, Unit Testing

Reviewer 2: Only use existing caterories!
category: Testing, Development, Frontend

comments: true
---

> # Javascript unit testing 101

Reviewer: ```suggestion
# JavaScript unit testing 101
```

> In this blog post I will explain how to write unit tests for javascript. For the examples, we will be using jasmine and ts-mockito, but the theory should be applicable to every testing framework. Why am I using two frameworks here instead of only jasmine? Well, in my opinion, jasmine relies too much on magic strings that might give problems during refactoring and makes autocompletion impossible in a lot of cases (depending on how good your IDE is). Ts-mockito is a mocking framework that allows you to create mocks based on a class, so you don’t have to tell it the method names to mock. There are other good mocking frameworks for typescript as well, but my experience is with ts-mockito, so I will be using it for the examples of this blog post. For the code examples I used the Angular CLI to generate a project for me, but I will not be using any of the Angular testing tools to keep these test cases completely framework independent. The Angular CLI was just an easy way to quickly get everything up and running.

Reviewer 3: JavaScript :)

 });
````

Then you write your verification part:

````javascript
it('should only return the cars where the color is black', () => {
      const actual = componentUnderTest.getAllBlackCars();

      expect(actual).toEqual([aBlackCadillac]);
      verify(carService.getCars()).once();
});
````

After this, you run your test once. It should fail. Then you write the setup you need to get to the expected output:

````javascript
it('should only return the cars where the color is black', () => {
      when(carService.getCars()).thenReturn([aBrownCar, aRedFerrari, aRedPorsche, aBlackCadillac]);

      const actual = componentUnderTest.getAllBlackCars();

      expect(actual).toEqual([aBlackCadillac]);
      verify(carService.getCars()).once();
});
````

This is a good way to work when you write unit tests. Writing the verification part before your setup, allows you to see that your test is actually able to fail. This is especially handy when you are testing asynchronous code, where a missing done operator can make your test be successful every time, no matter what your expects check, but more about this in the next part.

## Asynchronous testing

> Thinking in async code, can be a bit of a challenge, especially when you start out with javascript as a new developer or coming from a language that works mostly with synchronous code.

Reviewer: ```suggestion
Thinking in async code can be a bit of a challenge, especially when you start out with JavaScript as a new developer or coming from a language that works mostly with synchronous code.
```

In this part of the post, I will show you how to test asynchronous code. Both with promises and observables.

For both promises and observables, I will test methods that show the most common usage patterns. One scenario is where you simply get something asynchronously and set its value to a field, the other scenario is a little more complicated, where you get something asynchronously, have some side effects before it resolves, and have some more side effects after it resolves. If that seems complicated, don’t worry, it will become clear very soon.
For the promises, I will be using the async / await syntax. If you don’t know what that is yet, I suggest you read up on it, a good article is

> <a href="https://alligator.io/js/async-functions/">this one</a>.

Reviewer 2: try to provide your links like this:
```
[this one](https://alligator.io/js/async-functions/){:target="_blank" rel="noopener noreferrer"}.
```

### Promises

#### Scenario A

 The method we will be testing will be the following:
 
````javascript
async getAllRedCarsAsPromise(): Promise<Array<Car>> {
    const redCars = await this.carService.getCarsAsPromise();
    return redCars.filter((car: Car) => car.color === 'red');
}
````
 

> This function just gets the cars from the carService asynchronously and filters out all the non-red cars before returning the array. 

Reviewer: ```suggestion
This function just gets the cars from the `carService` asynchronously and filters out all the non-red cars before returning the array. 
```

> The first test, we will simply test the happy scenario, where the carService returns us a nice list of cars and the non-red cars are filtered out.

Reviewer: ```suggestion
The first test, we will simply test the happy scenario, where the `carService` returns us a nice list of cars and the non-red cars are filtered out.
```

````javascript
 describe('getAllRedCarsAsPromise', () => {

    it('should get the cars where the color is red', async done => {
      when(carService.getCarsAsPromise()).thenReturn(Promise.resolve([aRedPorsche, aBlackCadillac, aBrownCar]));

      try {
        const actual = await componentUnderTest.getAllRedCarsAsPromise();

        expect(actual).toEqual([aRedPorsche]);
      } catch (e) {
        fail(e);
      } finally {

>    done();

Reviewer 3: Indentation is missing?

 }
    });
});
````

> So what exactly goes on here? First we mock the carService.getCarsAsPromise() method and tell it that it should always return a promise that resolves with a list of cars. Then we call the method that we are testing, we wait for it to resolve (by using the await keyword). Afterward we expect that the returned array only contains the red Porsche. Now, what is this whole try catch thingy? When the promise is rejected, the execution of the code in the try block will be stopped and the code in the catch block will be executed. In this case, if the getCarsAsPromise() should fail, it will go into the catch block and the test will fail. In the finally block, we call the done function, to tell jasmine that our test is done, we call it in the finally block, because the code in the finally block will always be executed, regardless of the result of the promise. 

Reviewer 3: The test will fail, but what will `fail(e)` output? Handy for the user to already know they can also add a string as shown below in the next example?

````javascript
it('should set retrievalError to true if the promise is rejected', async done => {
      when(carService.getCarsAsPromise()).thenReturn(Promise.reject());

      try {
        await componentUnderTest.getAllRedCarsAsPromise();

        fail('Call should not have succeeded');
      } catch (e) {
        // nothing to expect here
      } finally {
        done();
      }
});
````

As you can see here, we call the fail function in the try block, because the method under test should throw an error. So after calling the method under test, it should go directly to the catch block. Since our method under test does not handle any errors, there is nothing to expect there. In the finally block, we tell jasmine the test is done.

#### Scenario B

The next method we will be testing is the following:

````javascript
async getAllRedCarsAsPromiseWithStuffHappeningInBetween(): Promise<void> {
    try {
      this.loading = true;
      this.cars = await this.getAllRedCarsAsPromise();
    } catch (e) {
      this.carRetrievalError = true;
    } finally {
      this.loading = false;
    }
  }
````

So a few things are happening:

> * First we set loading to true

Reviewer: ```suggestion
* First we set `loading` to true
```

> * Then we get the red cars asynchronously and assign them to our cars field

Reviewer: ```suggestion
* Then we get the red cars asynchronously and assign them to our `cars` field
```

> * If getting the red cars throws an error, we set carRetrievalError to true

Reviewer: ```suggestion
* If getting the red cars throws an error, we set `carRetrievalError` to `true`
```

> * Afterward we always set the loading back to false

Reviewer: ```suggestion
* Afterward we always set the `loading` back to false
```

````

So what’s happening here? In the promise callback, we assign the resolve function to a variable outside of that function, so we can use it later on. Then we call our method under test, but notice that we don’t use the await keyword here. Why is this? Well, if we were to use the await keyword, it would simply wait for the promise to resolve, but it would never happen because our resolve function is never called. Instead we just want to call it, let it execute and start expecting the side effects. As we saw in the method under test, the loading field is set to true before the red cars are retrieved, since those cars are not retrieved yet, we can check that the loading field is indeed set to true. Ok, now we have checked that, let's resolve the carService call. We do this by calling our promiseResolve function. Now the cars are resolved, so we can check that the loading has now been set to false and the cars field has been correctly set to the right value.
The finally callback of the returned promise, will execute after we call our promiseResolve function, and will check the final result of our method and subsequently call the done function to tell jasmine our test is over.

> In the second test we will test the error scenario, where we want to see if the loading value is still set correctly and the retrievalError is also set to true.

Reviewer: ```suggestion
In the second test we will test the error scenario, where we want to see if the loading value is still set correctly and the `carRetrievalError` is also set to true.
```

````javascript
it('should set the retrievalError to true and loading states correctly if the promise is rejected', async done => {
      let promiseReject: (cars: Array<Car>) => void;
      when(carService.getCarsAsPromise()).thenReturn(new Promise((resolve, reject) => {
        promiseReject = reject.bind(this);
      }));

      componentUnderTest
        .getAllRedCarsAsPromiseWithStuffHappeningInBetween()
        .finally(
          () => {
            expect(componentUnderTest.cars).toEqual([]);
            expect(componentUnderTest.loading).toBeFalsy();
            expect(componentUnderTest.carRetrievalError).toBeTruthy();
            done();
          }
        );

      expect(componentUnderTest.loading).toBeTruthy();

      promiseReject(undefined);
    });
````

> Here we actually do the same thing, except instead of using the resolve function, we use the reject function. 

Reviewer: ```suggestion
Here we actually do the same thing, except instead of using the resolve function, we use the `reject` function. 
```

### Observables

For observables, I have created methods that do exactly the same as our previous promises scenarios, except we use observables here.

#### Scenario A

>  Here we will test a simple observable that returns one value and then completes. This is a fairly common use case, especially with http calls.

Reviewer: ```suggestion
For Scenario A we will test a simple observable that returns one value and then completes. This is a fairly common use case, especially with http calls.
```
To avoid the double Here

 Here is our method under test:
 
````javascript
getAllRedCarsAsObservable(): Observable<Array<Car>> {
    return this.carService
      .getCarsAsObservable()
      .pipe(
        map((cars: Array<Car>) => cars.filter((car: Car) => car.color === 'red'))
       );
}
````
 

>  We simply get all the cars as an observable and then we call the map operator to filter the list we get. Pretty simple scenario.

Reviewer: ```suggestion
 We simply get all the cars as an observable and then we call the `map` operator to filter the list we get. Pretty simple scenario.
```

So our first test is the happy path, the observable resolves with an array of cars and everything goes as expected:

````javascript
it('should set all the red cars to the cars field and set loading field correctly', async done => {
    const carsSubject = new Subject<Array<Car>>();
    when(carService.getCarsAsObservable()).thenReturn(carsSubject);

    componentUnderTest.getAllRedCarsAsObservablewithStuffHappeningInBetween();

    expect(componentUnderTest.loading).toBeTruthy();

    carsSubject.next([aBlackCadillac, aRedPorsche]);
    carsSubject.complete();

    expect(componentUnderTest.cars).toEqual([aRedPorsche]);
    expect(componentUnderTest.loading).toBeFalsy();
    done();
});
````

This scenario is very easily tested with observables, because of Subjects. What is a subject? A subject can act as both an observable and an observer. This means we can subscribe to it and receive the emitted values, or we can tell it to emit a certain value.

> In this case, instead of telling the carService to return an Observable that immediately resolves with a value, we tell it to return a subject.asObservable(). Now we are in control of when and what value is emitted in this observable. This allows us to check our side effects easily at the right time.

Reviewer: ```suggestion
In this case, instead of telling the `carService` to return an observable that immediately resolves with a value, we tell it to return a `subject.asObservable()`. Now we are in control of when and what value is emitted in this observable. This allows us to check our side effects easily at the right time.
```

> Note that I am calling the carsSubject.complete() after I emit the car array value. I do this because I set the loading to false in the finalize operator. This finalize operator is only called when the observable is completed. So if I don’t call the carsSubject.complete(), the loading will never be set to false and my test will fail.

Reviewer: ```suggestion
Note that I am calling the `carsSubject.complete()` after I emit the car array value. I do this because I set the `loading` to `false` in the `finalize` operator. This finalize operator is only called when the observable is completed. So if I don’t call the `carsSubject.complete()`, the `loading` will never be set to `false` and my test will fail.
```
